Use same HTTP connection for multiple requests

anshul-patel-infostretch opened the pull request:
* Moved session initiation from client to CruxConfig.
* Moved requests out of [context manager](https://stackoverflow.com/questions/24873927/python-requests-module-and-connection-reuse).
* Added `close()` method to API client to manually close the session.
* Updated documentation.

## Test

* `make test` passes.
* Tested with the attached script to verify that requests are initiated via single session object. As  separate transport is used  for GCS `download` and `upload`, it will create separate connection for those requests.
* Verified the calls via the proxy logs and below program debug logs.

```python
from requests import Session
import logging
logging.basicConfig(level=logging.DEBUG)

from crux import Crux


proxies = {
    'http': 'http://localhost:3128',
    'https': 'http://localhost:3128'
}
conn = Crux(proxies=proxies)
conn.whoami()
do = conn.create_dataset("test_dataset_123")
fo = do.upload_file("/tmp/test.csv","/test.csv")
print(fo.id)
fo.download("/tmp/d_test.csv")
fo.delete()
do.delete()
conn.close()
input("Intentional block to see if connection gets closed before program termination")
```

antoncohen commented:
Ah, I had previously confused [`Session.close()`](http://docs.python-requests.org/en/master/api/#requests.Session.close) and [`Response.close()`](http://docs.python-requests.org/en/master/api/#requests.Response.close), and it is `Response.close()` then [needs to be called for streaming](http://docs.python-requests.org/en/master/user/advanced/#body-content-workflow). So it is good to see that added. Should we wrap those calls in a context manager to make sure `Response.close()` is called for streams, [like this](https://github.com/kennethreitz/requests/issues/4136)?

Does `Session.close()` need to be called? I'm thinking not. I took a look at [google-cloud-python](https://github.com/googleapis/google-cloud-python), it uses requests, reusing the session, and never calls `close()`. `google.cloud.storage.client.Client` has a [request session](https://github.com/googleapis/google-cloud-python/blob/master/storage/google/cloud/storage/client.py#L46), which [gets the session](https://github.com/googleapis/google-cloud-python/blob/master/core/google/cloud/client.py#L155) from `google.cloud.client.Client`, which gets it from `google.auth.transport.requests.AuthorizedSession`, which [uses a regular requests Session](https://github.com/googleapis/google-auth-library-python/blob/master/google/auth/transport/requests.py#L172). I searched the whole `google.cloud` codebase and as far as I can tell it never closes sessions when it [makes API requests](https://github.com/googleapis/google-cloud-python/blob/master/core/google/cloud/_http.py#L223).

anshul-patel-infostretch commented:
Thanks and Agreed @antoncohen , updated the code.

- Removed conn.close() from rest of the documentation.
- Added dedicated documentation for Crux Python Client.

anshul-patel-infostretch commented on docs/conf.py:
```diff
@@ -67,6 +67,9 @@ def setup(app):
 apidoc_output_dir = "./"
 
+# Add extra arguments which will be passed to sphinx-apidoc.
+apidoc_extra_args = ["-P"]
```
Hey @antoncohen , Added `-P` which will also publish documentation of private modules, for eg: `_client.py`, `compat.py`, `_utils.py`, etc. The reason is that it will help consuming development team in understanding the client. What do you think about it?

antoncohen replied:
I feel like the listing in [module documentation](https://crux-python.cruxinformatics.com/en/latest/modules.html) already has a lot. Overall I don't think private things like _utils should be publicly documented.

But I understand how having _client be documented makes sense. We are exposing it as a public attribute of `Crux`.

Do you think `Client` should be public? Are we comfortable with saying its interface will be stable?

My recommendation for this PR is to not publish docs for private modules. And then consider a future PR that makes `Client` public, if you think it should be.

anshul-patel-infostretch replied:
> Do you think `Client` should be public? Are we comfortable with saying its interface will be stable?
> 
> My recommendation for this PR is to not publish docs for private modules. And then consider a future PR that makes `Client` public, if you think it should be.

Ah, Got it and agreed, probably not now but may be in future we can wrap private module `_client.py` with public `client` module which will take care of compatibility safe to expose as public doc.

Thanks !
